data-pipeline/src/data_pipeline/pipelines/variant_cooccurrence_counts.py

```diff
@@ -0,0 +1,152 @@
+
+
+def build_counts(coordinates, counts, field_name_map):
+    struct_schema = {
+        processed_field_name: hl.or_else(
+            counts[coordinates.gene_id, coordinates.csq, coordinates.af_cutoff][raw_field_name], 0
+        )
+        for (processed_field_name, raw_field_name) in field_name_map.items()
+    }
+
+    return coordinates.annotate(counts=hl.struct(**struct_schema))
+
+
+def prepare_variant_cooccurrence_counts(tsv_path, field_name_map, csqs, af_cutoffs):
+    key_field_types = {"gene_symbol": hl.tstr, "csq": hl.tstr, "af_cutoff": hl.tstr}
+    input_field_types = dict(map(lambda field_name: (field_name, hl.tint), field_name_map.values()))
+
+    raw_counts = hl.import_table(
+        tsv_path,
+        types={**key_field_types, **input_field_types},
+        key=["gene_id", "csq"],
+        min_partitions=100,
+    )
+    raw_counts = raw_counts.transmute(af_cutoff=AF_CUTOFF_MAPPING_HL[raw_counts.af_cutoff])
+    raw_counts = raw_counts.key_by("gene_id", "csq", "af_cutoff")
+
+    gene_ids = raw_counts.key_by("gene_id").select()
+    gene_ids = gene_ids.distinct()
+    coordinates = generate_coordinates(gene_ids, csqs, af_cutoffs)
+
+    processed_counts = build_counts(coordinates, raw_counts, field_name_map)
+    processed_counts = processed_counts.group_by(processed_counts.gene_id, processed_counts.csq).aggregate(
+        counts_per_af=build_counts_per_af(processed_counts, af_cutoffs)
+    )
+    processed_counts = processed_counts.group_by(processed_counts.gene_id).aggregate(
+        variant_cooccurrence_counts=build_variant_cooccurrence_counts(processed_counts, csqs)
+    )
+    return processed_counts
+
+
+def prepare_heterozygous_variant_cooccurrence_counts():
+    field_name_map = {
+        "in_cis": "n_same_hap_without_chet_or_unphased",
+        "in_trans": "n_chet",
+        "unphased": "n_unphased_without_chet",
+        "two_het_total": "n_any_het_het",
+    }
+    return prepare_variant_cooccurrence_counts(
+        TWO_HET_DATA_PATH, field_name_map, HETEROZYGOUS_CSQS, HETEROZYGOUS_AF_CUTOFFS
+    )
+
+
+def prepare_homozygous_variant_cooccurrence_counts():
+    field_name_map = {"hom_total": "n_hom"}
+    return prepare_variant_cooccurrence_counts(
+        HOMOZYGOUS_DATA_PATH, field_name_map, HOMOZYGOUS_CSQS, HOMOZYGOUS_AF_CUTOFFS
+    )
+
+
+def annotate_table_with_variant_cooccurrence_counts(
+    genes_path=None,
+    heterozygous_variant_cooccurrence_counts_path=None,
+    homozygous_variant_cooccurrence_counts_path=None,
+):
+    genes = hl.read_table(genes_path)
+    heterozygous_variant_cooccurrence_counts = hl.read_table(heterozygous_variant_cooccurrence_counts_path)
+    homozygous_variant_cooccurrence_counts = hl.read_table(homozygous_variant_cooccurrence_counts_path)
+
+    # Note that we JSON-encode the counts before adding them to genes here.
+    # This is a workaround for a limitation of export_to_elasticsearch, which
+    # crashes if given a table with too many fields per row. These count
```

Comment:
Your many column implementation required JSON serialization... which to me is another indicator that this is not the best approach.